cmd/trtl/main.go

```diff
@@ -558,6 +575,95 @@ func initPeersClient(c *cli.Context) (err error) {
 // Trtl (DB) Client Functions
 //===========================================================================
 
+func migrateRecipients(c *cli.Context) (err error) {
+	ctx, cancel := profile.Context()
+	defer cancel()
+
+	// Iterate over the VASPs in the database
+	req := &pb.CursorRequest{
+		Namespace: wire.NamespaceVASPs,
+	}
+	var stream pb.Trtl_CursorClient
+	if stream, err = dbClient.Cursor(ctx, req); err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	for {
+		var rep *pb.KVPair
+		if rep, err = stream.Recv(); err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return cli.Exit(err, 1)
+		}
+
+		// Decode the VASP
+		vasp := &gds.VASP{}
+		if err = proto.Unmarshal(rep.Value, vasp); err != nil {
+			return cli.Exit(err, 1)
+		}
+
+		// Iterate over all contacts with an email address
+		var updated int
+		iter := models.NewContactIterator(vasp.Contacts, true, false)
+	contactsLoop:
+		for iter.Next() {
+			var entries int
+			contact, kind := iter.Value()
+
+			// Sanity check - the contact should have an email address
+			if contact.Email == "" {
+				return cli.Exit(fmt.Errorf("%s contact for vasp %s has no email address", kind, vasp.Id), 1)
+			}
+
+			// Get the extra data for the contact
+			extra := &models.GDSContactExtraData{}
+			if contact.Extra == nil {
+				continue contactsLoop
+			}
+
+			if err = contact.Extra.UnmarshalTo(extra); err != nil {
+				return cli.Exit(err, 1)
+			}
+
+			// Iterate over the email log and populate the recipients
+			for _, email := range extra.EmailLog {
+				if email.Recipient == "" {
+					email.Recipient = contact.Email
```

Comment:
Note that this might be inaccurate if the contact email was changed at some point (as older entries will contain the newer email address instead of the older address), we might have to fix that with another migration (or manually) if that ever happened.

Reply:
Interesting - this is certainly going to be true for several VASPs that have experienced staff turnover such as Anchorage, Apex Crypto, and probably a handful of others.  Ironically this would be a good use case for change data capture a la Ensign 😁  But I think we can live with slightly wrong data for some historical records, as the email log feature is primarily useful during the certificate issuance or reissuance process.

Reply:
Good to know, at least it will be accurate going forward.